src/omlt/neuralnet/nn_formulation.py

```diff
@@ -0,0 +1,264 @@
+
+def _ignore_input_layer():
+    pass
+
+_DEFAULT_LAYER_CONSTRAINTS = {
+    InputLayer: _ignore_input_layer,
+    DenseLayer: full_space_dense_layer,
+    ConvLayer: full_space_conv_layer,
+}
+
+_DEFAULT_ACTIVATION_CONSTRAINTS = {
+    "linear": linear_activation_constraint,
+    "relu": bigm_relu_activation_constraint,
+    "sigmoid": sigmoid_activation_constraint,
+    "softplus": softplus_activation_constraint,
+    "tanh": tanh_activation_constraint
+}
+
+class NeuralNetworkFormulation(_PyomoFormulation):
+    """
+    This class is the entry-point to build neural network formulations.
+
+    This class iterates over all nodes in the neural network and for
+    each one them, generates the constraints to represent the layer
+    and its activation function.
+
+    Parameters
+    ----------
+    network_structure : NetworkDefinition
+        the neural network definition
+    layer_constraints : dict-like or None
+        overrides the constraints generated for the specified layer types
+    activation_constraints : dict-like or None
+        overrides the constraints generated for the specified activation functions
+    """
+    def __init__(self, network_structure, layer_constraints=None, activation_constraints=None):
+        super().__init__()
+
+        self.__network_definition = network_structure
+        self.__scaling_object = network_structure.scaling_object
+        self.__scaled_input_bounds = network_structure.scaled_input_bounds
+        
+        if layer_constraints is None:
+            layer_constraints = dict()
+        if activation_constraints is None:
+            activation_constraints = dict()
+
+        self._layer_constraints = {**_DEFAULT_LAYER_CONSTRAINTS, **layer_constraints}
+        self._activation_constraints = {**_DEFAULT_ACTIVATION_CONSTRAINTS, **activation_constraints}
+
+    def _build_formulation(self):
+        _setup_scaled_inputs_outputs(self.block,
+                                     self.__scaling_object,
+                                     self.__scaled_input_bounds)
+        
+        _build_neural_network_formulation(
+            block=self.block,
+            network_structure=self.__network_definition,
+            layer_constraints=self._layer_constraints,
+            activation_constraints=self._activation_constraints,
+        )
+
+    @property
+    def input_indexes(self):
+        """The indexes of the formulation inputs."""
+        network_inputs = list(self.__network_definition.input_nodes)
+        assert len(network_inputs) == 1, 'Unsupported multiple network input variables'
+        return network_inputs[0].input_indexes
+
+    @property
+    def output_indexes(self):
+        """The indexes of the formulation output."""
+        network_outputs = list(self.__network_definition.output_nodes)
+        assert len(network_outputs) == 1, 'Unsupported multiple network output variables'
+        return network_outputs[0].output_indexes
```

Comment:
I would have these errors say something akin to `unsupported multiple layers` versus variables

Reply:
Fixed text and added these to __init__ so we find out sooner.